pkg/operation/botanist/etcdencryption.go

```diff
@@ -0,0 +1,321 @@
+	enableEtcdEncryption, err := utils.CheckVersionMeetsConstraint(b.Shoot.Info.Spec.Kubernetes.Version, ">= 1.13")
+	if err != nil {
+		return err
+	}
+	if !enableEtcdEncryption {
+		return nil
+	}
+	needToWriteConfig := false
+	exists, ecSeed, forcePlaintextSecrets, err := b.readEncryptionConfigurationFromSeed()
+	if err != nil {
+		return err
+	}
+	if !exists {
+		// create new passive EncryptionConfiguration if it does not exist yet
+		// and remember to write this created configuration
+		b.Logger.Info("no EncryptionConfiguration found in seed. Creating new passive configuration.")
+		ecSeed, err = encryptionconfiguration.CreateNewPassiveConfiguration()
+		if err != nil {
+			return err
+		}
+		needToWriteConfig = true
+	} else {
+		exists, ecGarden, err := b.readEncryptionConfigurationFromGarden()
+		if err != nil {
+			return err
+		}
+		if !exists {
+			return fmt.Errorf("EncryptionConfiguration exists in seed but not in garden")
+		}
+		// if it exists already, it needs to be consistent
+		consistent, err := b.isEncryptionConfigurationConsistent(ecSeed, ecGarden)
+		if err != nil {
+			return err
+		}
+		if !consistent {
+			return fmt.Errorf("EncryptionConfiguration in seed is not consistent")
+		}
+		if forcePlaintextSecrets {
+			if encryptionconfiguration.IsActive(ecSeed) {
+				// if annotation requires plaintext secrets, deactivate configuration if it is still active
+				b.Logger.Infof("EncryptionConfiguration is annotated with %v. De-activating existing active EncryptionConfiguration. Note: shoot secrets will be decrypted in etcd subsequently", common.EtcdEncryptionForcePlaintextAnnotationName)
+				encryptionconfiguration.SetActive(ecSeed, false)
+				needToWriteConfig = true
+			}
+		} else if !encryptionconfiguration.IsActive(ecSeed) {
+			// if it is not active already (meaning that identity and not aescbc
+			// is still the first encryption provider) then set it to active
+			// and remember to write this created configuration
+			b.Logger.Info("activating existing passive EncryptionConfiguration")
+			encryptionconfiguration.SetActive(ecSeed, true)
+			needToWriteConfig = true
+		}
+	}
+	ecSeedYamlBytes, err := encryptionconfiguration.ToYAML(ecSeed)
+	if err != nil {
+		return err
+	}
+	if needToWriteConfig {
+		b.Logger.Info("writing new/updated EncryptionConfiguration")
+		if err := b.writeEncryptionConfiguration(ecSeedYamlBytes, forcePlaintextSecrets); err != nil {
+			return err
+		}
+	}
+
+	checksum := utils.ComputeSHA256Hex(ecSeedYamlBytes)
+	b.mutex.Lock()
+	b.CheckSums[common.EtcdEncryptionSecretName] = checksum
+	b.mutex.Unlock()
+	// enablement of etcd encryption feature done in helm chart of apiserver deployment
+	return nil
+}
+
+// RewriteShootSecretsIfEncryptionConfigurationChanged rewrites a shoot's secrets if the EncryptionConfiguration has changed
+func (b *Botanist) RewriteShootSecretsIfEncryptionConfigurationChanged(ctx context.Context) error {
```

Comment:
The configuration is written before the apiserver is updated since the configuration is required for its start (and if the configuration is changed, the apiserver needs to get restarted). 

The rewriting of secrets is done much later only once the apiserver becomes available with the correct configuration. 

I do not understand how these two things can me merged in a single function (with the apiserver restart in between) ...

Reply:
You could e.g. set an annotation on a secret once it's rewritten - then you'd also not need to send an update for already up-to-date secrets. This could be done in such a function.

Reply:
The rewriting of secrets happens once for the shoot cluster ... and if it is successful, it will never occur again (as long as the EncryptionConfiguration does not change ... and it does not change unless key rotation is implemented here, which has explicitly not been requested). 

I am not sure how often it occurs that the retry.RetryOnConflict(retry.DefaultRetry .. will fail for a secret or if this is rather a case that should be investigated anyway if it failed. If this is rather unusual, I think it would be better not to annotate each customer secret (e.g. with the checksum of the key that the secret is encrypted with).

Of course if you tell me that it is the normal case that the 5 retries to update a secret all fail in gardener, then I agree that it is better to remember the secrets for which the rewrite worked in order not to (unnecessarily) rewrite them during the next reconciliation.

Reply:
Consider the case that a shoot has 10000 secrets and we need to rewrite. During the process of rewriting the network connectivity might break, and rewriting 10000 secrets is taking quite some time.

So, when network works again we don't want/need to rewrite the secrets that already succeeded to get rewritten, and in this situation we would benefit from a label that we put to secrets in order to filter the already updated secrets.